### PR TITLE
Speed up synthetic source

### PR DESCRIPTION
**Pull request by nik9000**
This speeds up synthetic source, especially when there are many fields
in the index that are declared in the mapping but don't have values.
This is fairly common with ECS, and the tsdb rally track uses that. And
this improves fetch performance of that track:
```
|  50th percentile service time |    default |   6.24029 |  4.85568 | ms | -22.19% |
|  90th percentile service time |    default |   7.89923 |  6.52069 | ms | -17.45% |
|  99th percentile service time |    default |  12.0306  | 16.435   | ms | +36.61% |
| 100th percentile service time |    default |  14.2873  | 17.1175  | ms | +19.81% |
|  50th percentile service time | default_1k | 158.425   | 25.3236  | ms | -84.02% |
|  90th percentile service time | default_1k | 165.46    | 30.8655  | ms | -81.35% |
|  99th percentile service time | default_1k | 168.954   | 33.3342  | ms | -80.27% |
| 100th percentile service time | default_1k | 174.341   | 34.8344  | ms | -80.02% |
```

There's a slight increase in the 99th and 100th percentile service time
for fetching ten document which think is unlucky jitter. Hopefully. The
average performance of fetching ten docs improves anyway so I think
we're ok. Fetching a thousand documents improves 80% across the board
which is lovely.

This works by doing three things:
1. Teach the "leaf" layer of source loader to detect when the field is
   empty in that segment and remove it from the synthesis process
   entirely. This brings most of the speed up in tsdb.
2. Replace `hasValue` with `advanceToDoc` returning `boolean` and
    cache the result.
3. Replace the `ArrayList` of leaf loaders with an array. Before fixing
   the other two issues the `ArrayList`'s iterator really showed up in
   the profiling. Probably much less worth it now, but it's small.

All of this brings synthetic source much closer to the fetch performance
of standard _source:
```
|  50th percentile service time | default_1k |  11.4016  | 25.3236  | ms | +122.11% |
|  90th percentile service time | default_1k |  13.7212  | 30.8655  | ms | +124.95% |
|  99th percentile service time | default_1k |  15.8785  | 33.3342  | ms | +109.93% |
| 100th percentile service time | default_1k |  16.9715  | 34.8344  | ms | +105.25% |
```

One important thing, these perf numbers come from fetching *hot* blocks
on disk. They mostly compare CPU overhead and not disk overhead.

image

### CODE REVIEW

**Comment by nik9000** [comment]
labelbot I have set a label

File: server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java [hunk 1]
```diff
@@ -1629,24 +1631,43 @@ protected NumericSyntheticFieldLoader(String name, String simpleName) {
...
+            SortedNumericDocValues dv = reader.getSortedNumericDocValues(name);
+            if (dv != null) {
+                return dv;
+            }
+            NumericDocValues single = reader.getNumericDocValues(name);
+            if (single != null) {
+                return DocValues.singleton(single);
+            }
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:1644]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 2]
```diff
@@ -554,51 +555,56 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
...
                     @Override
-                    public void load(XContentBuilder b) throws IOException {
-                        boolean started = false;
-                        for (SourceLoader.SyntheticFieldLoader.Leaf leaf : leaves) {
-                            if (leaf.hasValue()) {
-                                if (false == started) {
-                                    started = true;
-                                    startSyntheticField(b);
+                    public void load(XContentBuilder b, CheckedRunnable<IOException> before) throws IOException {
+                        class HasValue implements CheckedRunnable<IOException> {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:590]: > Let me have a look. I thought it wouldn't be man I would love to get rid of this thing.

I... can't.... So close....

**Comment by nik9000** [comment]
@romseygeek this is ready for you again any time!

File: server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java [hunk 3]
```diff
@@ -82,14 +82,24 @@ public boolean reordersFieldValues() {
         @Override
         public Leaf leaf(LeafReader reader) throws IOException {
             SyntheticFieldLoader.Leaf leaf = loader.leaf(reader);
+            if (leaf.empty()) {
+                return new Leaf() {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java:86]: Given this has no state I wonder if it's worth having it as a final instance on Leaf?

**Review by romseygeek — APPROVED** [review]
Thanks, this looks much nicer! A couple of nits but LGTM otherwise, no need for another review.

File: server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java (hunk 3 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java:86]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java [hunk 4]
```diff
@@ -104,20 +114,19 @@ public BytesReference source(FieldsVisitor fieldsVisitor, int docId) throws IOEx
      * Load a field for {@link Synthetic}.
      */
     interface SyntheticFieldLoader {
-        /**
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java:107]: Not sure what I did there.

**Pull request merged**